src/conn.c

```diff
@@ -18,34 +19,45 @@ static int init_read(struct conn *c, uv_buf_t *buf, size_t size)
 }
 
 static int read_message(struct conn *c);
-static void conn_write_cb(struct transport *transport, int status)
+static void conn_write_cb(uv_write_t *req, int status)
 {
-	struct conn *c = transport->data;
+	struct transport *t = req->data;
+	assert(t != NULL);
+	struct conn *c = t->data;
+	assert(c != NULL);
 	bool finished;
 	int rv;
 	if (status != 0) {
 		tracef("write cb status %d", status);
 		goto abort;
 	}
+	if (c->closed) {
+		tracef("connection closing");
+		goto abort;
+	}
 
 	buffer__reset(&c->write);
 	buffer__advance(&c->write, message__sizeof(&c->response)); /* Header */
 
 	rv = gateway__resume(&c->gateway, &finished);
+	tracef("request finished: %d", finished);
 	if (rv != 0) {
 		goto abort;
 	}
-	if (!finished) {
-		return;
-	}
 
-	/* Start reading the next request */
+	/* Start reading the next message if we're not doing that already. */
+	if (c->await_message) {
+		goto exit;
+	}
 	rv = read_message(c);
 	if (rv != 0) {
 		goto abort;
 	}
+exit:
+	free(req);
 	return;
```

Comment:
since it's the only use of this label I think the control flow would be clearer if you just inlined `free(req); return` above